File: src/main/java/net/draconia/askaround/domain/AddressLine.java
package net.draconia.askaround.domain;

import java.io.Serializable;

import org.codehaus.jackson.annotate.JsonIgnore;

public class AddressLine implements Comparable<AddressLine>, Serializable
{
	private static final long serialVersionUID = 7734615744636182703L;
	
	private Address mObjAddress;
	private Integer miId, miLineNumber;
	private String msAddressLine;
	
	public AddressLine()
	{ }
	
	public AddressLine(final Address objAddress)
	{
		setAddress(objAddress);
	}
	
	public AddressLine(final Integer iLineNumber, final String sAddressLine)
	{
		setLineNumber(iLineNumber);
		setAddressLine(sAddressLine);
	}
	
	public AddressLine(final Address objAddress, final Integer iLineNumber, final String sAddressLine)
	{
		setAddress(objAddress);
		setLineNumber(iLineNumber);
		setAddressLine(sAddressLine);
	}
	
	public AddressLine(final Integer iId, final Integer iLineNumber, final String sAddressLine)
	{
		setId(iId);
		setLineNumber(iLineNumber);
		setAddressLine(sAddressLine);
	}
	
	public AddressLine(final Address objAddress, final Integer iId, final Integer iLineNumber, final String sAddressLine)
	{
		setAddress(objAddress);
		setId(iId);
		setLineNumber(iLineNumber);
		setAddressLine(sAddressLine);
	}
	
	public int compareTo(final AddressLine objAddressLine)
	{
		return(new Integer(getLineNumber()).compareTo(objAddressLine.getLineNumber()));
	}
	
	public boolean equals(final Object objAddressLine)
	{
		if((objAddressLine instanceof AddressLine) && (objAddressLine != null))
			{
			AddressLine objLine = ((AddressLine)(objAddressLine));
			
			if((objLine.getAddress() == null) && (objLine.getAddress() == getAddress()))
				return(new Integer(objLine.getLineNumber()).equals(getLineNumber()) && objLine.getAddressLine().equals(getAddressLine()));
			else if((getAddress() != null) && (getAddress().equals(objLine.getAddress())))
				return(new Integer(objLine.getLineNumber()).equals(getLineNumber()) && objLine.getAddressLine().equals(getAddressLine()));
			else
				return(false);
			}
		else
			return(false);
	}
	
	@JsonIgnore
	public Address getAddress()
	{
		return(mObjAddress);
	}
	
	public String getAddressLine()
	{
		if(msAddressLine == null)
			msAddressLine = "";
		
		return(msAddressLine);
	}
	
	public int getId()
	{
		if(miId == null)
			miId = 0;
		
		return(miId);
	}
	
	public int getLineNumber()
	{
		if(miLineNumber == null)
			miLineNumber = 0;
		
		return(miLineNumber);
	}
	
	public void setAddress(final Address objAddress)
	{
		mObjAddress = objAddress;
	}
	
	public void setAddressLine(final String sAddressLine)
	{
		if(sAddressLine == null)
			msAddressLine = "";
		else
			msAddressLine = sAddressLine;
	}
	
	public void setId(final Integer iId)
	{
		if(iId == null)
			miId = 0;
		else
			miId = iId;
	}
	
	public void setLineNumber(final Integer iLineNumber)
	{
		if(iLineNumber == null)
			miLineNumber = 0;
		else
			miLineNumber = iLineNumber;
	}
}

File: src/main/webapp/scripts/models/addressline.js
function AddressLine(objAddress, iId, sAddressLine, iLineNumber)
{
	this.mObjAddress					=	objAddress;
	this.msAddressLine					=	Common.defaultIfUndefined(sAddressLine, "");
	this.mbChanged						=	undefined;
	this.miId							=	Common.defaultIfUndefined(iId, 0);
	this.miLineNumber					=	Common.defaultIfUndefined(iLineNumber, 0);
	this.mArrObjObservers				=	[];
}

AddressLine.prototype.addObserver				=	function(objObserver)
{
	if(objObserver != undefined)
		{
		if(this.mArrObjObservers == undefined)
			this.mArrObjObservers = [];
		
		this.mArrObjObservers.push(objObserver);
		}
	
	return(this);
};

AddressLine.copyFromJSON						=	function(objAddress, objJSON, objAddressLine)
{
	if(objAddressLine == undefined)
		objAddressLine							=	new AddressLine(objAddress);
	
	for(prop in objJSON)
		objAddressLine["set" + prop.substring(0, 1).toUpperCase() + prop.substring(1)].call(objAddressLine, eval("objJSON." + prop));
	
	return(objAddressLine);
};

AddressLine.prototype.getAddress		=	function()
{
	return(this.mObjAddress);
};

AddressLine.prototype.getAddressLine	=	function()
{
	if(this.msAddressLine == undefined)
		this.msAddressLine = "";
	
	return(this.msAddressLine);
};

AddressLine.prototype.getId				=	function()
{
	if(this.miId == undefined)
		this.miId = 0;
	
	return(this.miId);
};

AddressLine.prototype.getLineNumber		=	function()
{
	if(this.miLineNumber == undefined)
		this.miLineNumber = 0;
	
	return(this.miLineNumber);
};

AddressLine.prototype.hasChanged				=	function()
{
	if(this.mbChanged == undefined)
		this.mbChanged = false;
	
	return(this.mbChanged);
};

AddressLine.prototype.notifyObservers			=	function(objArgument)
{
	this.mbChanged = true;
	
	for(var iLength = this.mArrObjObservers.length, iLoop = 0; iLoop < iLength; iLoop++)
		this.mArrObjObservers[iLoop].call(this, objArgument);
	
	this.mbChanged = false;
};

AddressLine.prototype.setAddress		=	function(objAddress)
{
	this.mObjAddress 					=	objAddress;
	
	return(this);
};

AddressLine.prototype.setAddressLine	=	function(sAddressLine)
{
	if(sAddressLine == undefined)
		this.msAddressLine				=	"";
	else
		this.msAddressLine 				=	sAddressLine;
	
	return(this);
};

AddressLine.prototype.setId				=	function(iId)
{
	if(iId == undefined)
		this.miId 						=	0;
	else
		this.miId						=	iId;
	
	return(this);
};

AddressLine.prototype.setLineNumber		=	function(iLineNumber)
{
	if(iLineNumber == undefined)
		this.miLineNumber				=	0;
	else
		this.miLineNumber 				=	iLineNumber;
	
	return(this);
};